Mathlib/Data/QuotLike.lean

```diff
@@ -0,0 +1,666 @@
+  let v ← match ← inferType Q with | .sort v => pure v | _ => mkFreshLevelMVar
+  have Q : Q(Sort v) := Q
+  let α ← mkFreshExprMVarQ q(Sort $(← mkFreshLevelMVar))
+  let r ← mkFreshExprMVarQ q($α → $α → Prop)
+  let .some inst ← trySynthInstanceQ q(QuotLike $Q $α $r) |
+    tryPostpone
+    throwError "Cannot find `QuotLike` instance for type `{Q}`."
+  pure q(@mkQ $Q $α $r $inst $(← Qq.elabTermEnsuringTypeQ a q($α)))
+
+open PrettyPrinter.Delaborator SubExpr in
+/-- Delaborator for `mkQ` -/
+@[delab app.QuotLikeStruct.mkQ]
+def delabMkQ : Delab := do
+  guard <| (← getExpr).isAppOfArity' ``mkQ 5
+  let a ← withNaryArg 4 delab
+  `(⟦$a⟧)
+
+@[inherit_doc mkQ]
+macro "⟦" a:term " : " α:term "⟧" : term => `(⟦($a : $α)⟧)
+
+/--
+`QuotLike.HasQuot` is used for deriving `QuotLike` instances from the input type.
+
+The instances of `QuotLike.HasQuot` may be defined for types that have specific uses.
+
+```
+instance : QuotLike.HasQuot ZFSet PSet PSet.Equiv where
+```
+
+They are also usually defined as scoped or local instances.
+
+```
+scoped instance {α} [s : Setoid α] : QuotLike.HasQuot (Quotient s) α (· ≈ ·) where
+```
+-/
+class HasQuot (Q : outParam Sort*) (α : Sort*) (r : outParam (α → α → Prop))
+    [QuotLike Q α r] : Prop where
+
+/-- The canonical quotient map. Inferred from the input type via typeclass `QuotLike.HasQuot`. -/
+syntax (name := mkQ') "mkQ'" : term
+
+@[term_elab QuotLike.mkQ', inherit_doc QuotLike.mkQ']
+def mkQ'Impl : TermElab := fun stx typ? => do
+  let .some expectedType := typ? |
+    let α ← mkFreshTypeMVar
+    let β ← mkFreshTypeMVar
+    postponeElabTerm stx (some (← mkArrow α β))
+  synthesizeSyntheticMVars
+  let expectedType ← instantiateMVars expectedType
+  let expectedType ← whnf expectedType
+  let .forallE _ α _ _ := expectedType |
+    if expectedType.isMVar then tryPostpone
+    throwError "Expected type is not a function."
+  if α.isMVar then
+    tryPostpone
+    throwError "The input type must be known."
+  let Q ← mkFreshExprMVarQ q(Sort $(← mkFreshLevelMVar))
+  let u ← match ← inferType α with | .sort u => pure u | _ => mkFreshLevelMVar
+  have α : Q(Sort u) := α
+  let r ← mkFreshExprMVarQ q($α → $α → Prop)
+  let inst ← mkFreshExprMVarQ q(QuotLike $Q $α $r)
+  let .some _ ← trySynthInstanceQ q(@HasQuot $Q $α $r $inst) |
+    tryPostpone
+    throwError "Cannot find `QuotLike.HasQuot` instance for type `{α}`."
+  pure q(@mkQ $Q $α $r $inst)
+
+/-- The canonical quotient map. Inferred from the input type. -/
+macro "⟦" t:term "⟧'" : term => `(mkQ' $t)
+
+/-- The canonical quotient map. Inferred from the input type. -/
+macro "⟦" t:term " : " α:term "⟧'" : term => `(⟦($t : $α)⟧')
+
+/--
+`QuotLike.Hint` is used for deriving `QuotLike` instances from the hint.
+
+```
+scoped instance {α} (r : α → α → Prop) : QuotLike.Hint r (Quot r) α r where
+```
+
+```
+instance (p : Submodule R M) : QuotLike.Hint p (M ⧸ p) M p.quotientRel where
+```
+-/
+class Hint {Hint : Sort*} (hint : Hint)
```

Comment:
I don't understand, what is the hint used for, and how is it used?

Reply:
This is an interesting design. I do wonder if we could get away with just writing `(⟦x⟧ : M ⧸ p)`, though.